src/main/java/com/google/sps/database/DatabaseImpl.java

```diff
@@ -0,0 +1,190 @@
+    public int addClauses(String videoID, Map<String, List<Long>> clauses) {
+        for (Map.Entry<String, List<Long>> me : clauses.entrySet()) {
+            addClause(videoID, me.getKey(), me.getValue());  //error possibility
+        }
+
+        return 0;
+    }
+    
+    // add metadata information to a particular video in the db
+    public int addMetadata(String videoID, String metadata, boolean overwrite) {
+        Entity vidEnt = getVideo(videoID); //error possibility
+        Entity metaEnt;
+        
+        // need to find metadata already in the database
+        Entity currMeta = getMetadata(videoID);  //error possibility
+        Key metaKey = currMeta.getKey();  //error possibility
+
+        if (overwrite) {
+            datastore.delete(metaKey); //error possibility
+            metaEnt = new Entity("metadata", metadata, vidEnt.getKey()); //error possibility
+            datastore.put(metaEnt); //error possibility
+        } else {
+            /*
+            try {
+                metaEnt = datastore.get(metaKey); //error possibility
+            } catch (Exception e) {
+                return -1;
+            }
+            String current = metaEnt.getProperty("ID/Name").toString(); //error possibility
+            datastore.delete(metaKey); //error possibility
+            metaEnt = new Entity("metadata", current + metadata, vidEnt.getKey()); //error possibility
+            */
+            
+        }
+        //datastore.put(metaEnt); //error possibility
+
+        return 0;
+    }
+    
+    // retrieve all keywords + their timestamps in a specified videoID
+    public Map<String, List<Long>> getAllKeywords(String videoID) {
+        Entity vidEnt = getVideo(videoID);  //error possibility
+        Key vidKey = vidEnt.getKey();  //error possibility
+
+        Query query = new Query("caption", vidKey); //error possibility
+        PreparedQuery results = datastore.prepare(query); //error possibility
+
+        Map<String, List<Long>> hm = new HashMap<String, List<Long>>();
+
+        for (Entity e : results.asIterable()) {
+            hm.put(e.getKey().getName(), (List<Long>)e.getProperty("timestamps")); //error possibility
+        }
+
+        return hm;
+    }
+    
+    // retrieve specific timestamps for a specified keyword belonging to videoID
+    public List<Long> getTimesForKeyword(String videoID, String keyword) {
+        Map<String, List<Long>> hm = getAllKeywords(videoID); //error possibility
+
+        return hm.get(keyword); //error possibility
+    }
+
+    // return true if specified video is in the database
+    public boolean videoInDb(String videoID) {
+        if (getVideo(videoID) == null) return false;
+        else return true;
+    }
+
+    //================================================================================
+    // Private Helper Functions
+    //================================================================================
```

Comment:
It seems that the private helper functions are not functions in the interface. If this was intentional, add method header comments to these functions. Else, add the methods to the interface.